Question 1.1:

In [27]:
import numpy as np
import matplotlib.pyplot as plt
import mltools as ml
import math as math
entropy_y = 0.6*math.log(10/6.0,2) + 0.4*math.log(10/4.0,2)
print("entropy H(y)=",entropy_y)

entropy H(y)= 0.9709505944546687


Question 1.2:

In [30]:
# x1 information gain 
entropy_1_1 = (3.0/6)*math.log(6.0/3,2) + (3.0/6)*math.log(6.0/3,2)
entropy_1_0 = (3.0/4)*math.log(4.0/3,2) + (1.0/4)*math.log(4.0,2)
information_gain_1 = (6.0/10)*(entropy_y - entropy_1_1) + (4.0/10)*(entropy_y - entropy_1_0)
print('Information gain for feature 1:, %0.4f' %(information_gain_1))

# x2 information gain 
ent_2_1 = (5.0/5)*math.log(5.0/5,2)
ent_2_0 = (4.0/5)*math.log(5.0/4,2) + (1.0/5)*math.log(5.0,2)
information_gain_2 = (5.0/10)*(entropy_y - ent_2_1) + (5.0/10)*(entropy_y - ent_2_0)
print('Information gain for feature 2:, %0.4f' %(information_gain_2))

# x3 information gain 
ent_3_1 = (3.0/7)*math.log(7.0/3,2) + (4.0/7)*math.log(7.0/4,2)
ent_3_0 = (1.0/3)*math.log(3.0,2) + (2.0/3)*math.log(3.0/2,2)
information_gain_3 = (7.0/10)*(entropy_y - ent_3_1) + (3.0/10)*(entropy_y - ent_3_0)
print('Information gain for feature 3:, %0.4f' %(information_gain_3))


Information gain for feature 1:, 0.0464
Information gain for feature 2:, 0.6100
Information gain for feature 3:, 0.0058


In [31]:
# x4 information gain 
ent_4_1 = (2.0/7)*math.log(7.0/2,2) + (5.0/7)*math.log(7.0/5,2)
ent_4_0 = (2.0/3)*math.log(3.0/2,2) + (1.0/3)*math.log(3.0,2)
information_gain_4 = (7.0/10)*(entropy_y - ent_4_1) + (3.0/10)*(entropy_y - ent_4_0)
print('Information gain for feature 4:, %0.4f' %(information_gain_4))

# x5 information gain 
ent_5_1 = (1.0/3)*math.log(3.0,2) + (2.0/3)*math.log(3.0/2,2)
ent_5_0 = (3.0/7)*math.log(7.0/3,2) + (4.0/7)*math.log(7.0/4,2)
information_gain_5 = (3.0/10)*(entropy_y - ent_5_1) + (7.0/10)*(entropy_y - ent_5_0)
print('Information gain for feature 5:, %0.4f' %(information_gain_5))

Information gain for feature 4:, 0.0913
Information gain for feature 5:, 0.0058


I should split on variable 2 for the root node of the decision tree since it has the highest Information gain 0.6100

Question 1.3:
<img src="q1.3.png" alt="Drawing" style="width:550px;height:340px"/>

Question 2.1:

In [29]:
xt = np.genfromtxt('data/X_train.txt', delimiter=None)
yt = np.genfromtxt('data/Y_train.txt', delimiter=None)
xt,yt = ml.shuffleData(xt,yt)
for i in range(X.shape[1]):
    print('minimum of x%d:, %0.4f' %(i,min(xt[:,0])))
    print('maximum of x%d:, %0.4f' %(i,max(xt[:,0])))
    print('mean of x%d:, %0.4f' %(i,np.mean(xt[:,0])))
    print('mean of x%d:, %0.4f' %(i,np.var(xt[:,0])))
    print()

minimum of x0:, 197.0000
maximum of x0:, 253.0000
mean of x0:, 241.8990
mean of x0:, 81.1988

minimum of x1:, 197.0000
maximum of x1:, 253.0000
mean of x1:, 241.8990
mean of x1:, 81.1988

minimum of x2:, 197.0000
maximum of x2:, 253.0000
mean of x2:, 241.8990
mean of x2:, 81.1988

minimum of x3:, 197.0000
maximum of x3:, 253.0000
mean of x3:, 241.8990
mean of x3:, 81.1988

minimum of x4:, 197.0000
maximum of x4:, 253.0000
mean of x4:, 241.8990
mean of x4:, 81.1988

minimum of x5:, 197.0000
maximum of x5:, 253.0000
mean of x5:, 241.8990
mean of x5:, 81.1988

minimum of x6:, 197.0000
maximum of x6:, 253.0000
mean of x6:, 241.8990
mean of x6:, 81.1988

minimum of x7:, 197.0000
maximum of x7:, 253.0000
mean of x7:, 241.8990
mean of x7:, 81.1988

minimum of x8:, 197.0000
maximum of x8:, 253.0000
mean of x8:, 241.8990
mean of x8:, 81.1988

minimum of x9:, 197.0000
maximum of x9:, 253.0000
mean of x9:, 241.8990
mean of x9:, 81.1988

minimum of x10:, 197.0000
maximum of x10:, 253.0000
mean of 